pkg/persistentvolumeclaim/cstor.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2020-2021 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package persistentvolumeclaim
+
+import (
+	"fmt"
+	"time"
+
+	cstortypes "github.com/openebs/api/v2/pkg/apis/types"
+	"github.com/openebs/openebsctl/pkg/client"
+	"github.com/openebs/openebsctl/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/printers"
+)
+
+const (
+	cstorPvcInfoTemplate = `
+{{.Name}} Details :
+------------------
+NAME             : {{.Name}}
+NAMESPACE        : {{.Namespace}}
+CAS TYPE         : {{.CasType}}
+BOUND VOLUME     : {{.BoundVolume}}
+ATTACHED TO NODE : {{.AttachedToNode}}
+POOL             : {{.Pool}}
+STORAGE CLASS    : {{.StorageClassName}}
+SIZE             : {{.Size}}
+USED             : {{.Used}}
+CV STATUS	 : {{.CVStatus}}
+PV STATUS        : {{.PVStatus}}
+`
+
+	detailsFromCVC = `
+Additional Details from CVC :
+-----------------------------
+NAME          : {{ .metadata.name }}
+REPLICA COUNT : {{ .spec.provision.replicaCount }}
+POOL INFO     : {{ .status.poolInfo}}
+VERSION       : {{ .versionDetails.status.current}}
+UPGRADING     : {{if eq .versionDetails.status.current .versionDetails.desired}}false{{else}}true{{end}}
+`
+)
+
+// DescribeCstorVolumeClaim describes a cstor storage engine PersistentVolumeClaim
+func DescribeCstorVolumeClaim(c *client.K8sClient, pvc *corev1.PersistentVolumeClaim, pv *corev1.PersistentVolume) error {
+	// Create Empty template objects and fill gradually when underlying sub CRs are identified.
+	pvcInfo := util.CstorPVCInfo{}
+
+	pvcInfo.Name = pvc.Name
+	pvcInfo.Namespace = pvc.Namespace
+	pvcInfo.BoundVolume = pvc.Spec.VolumeName
+	pvcInfo.CasType = util.CstorCasType
+	pvcInfo.StorageClassName = *pvc.Spec.StorageClassName
+
+	if pv != nil {
+		pvcInfo.PVStatus = pv.Status.Phase
+	}
+
+	// fetching the underlying CStorVolume for the PV, to get the phase and size and notify the user
+	// if the CStorVolume is not found.
+	cv, err := c.GetCV(pvc.Spec.VolumeName)
+	if err != nil {
+		fmt.Println("Underlying CstorVolume is not found for: ", pvc.Name)
```

Comment:
So there can be several reasons:
- if CVC does not exists then something might be wrong with csi plugin pods (controller & node) / wrong configuration in SC / kube-apiserver not reachable / etc. It can be debugged by looking at controller and node logs.
- if only CV is missing then something something wrong with the CVC produced by the csi node, the version on CVC does not match the CVC operator version / CVC operator is down due to some reason / wrong config in the CVC / etc. This can be debugged by looking at CVC operator logs

Reply:
For now the current code is good. We should just return the error and we can improve on debugging in later PRs

Reply:
This is great, I've kept a backup of this info #42

On implementation side,
1. Okay, so some potential red-flags would be `>= 1 restartCount` for CSI pods & `some events thrown around in that namespace`.
2. Is this valid only for CStor or is it a general OpenEBS practice to avoid issues?